forc-plugins/forc-doc/src/render/item/context.rs

```diff
@@ -0,0 +1,354 @@
+use crate::{
+    doc::module::ModuleInfo,
+    render::{
+        constant::IDENTITY, item::type_anchor::render_type_anchor, link::*, title::DocBlockTitle,
+        title::*, util::format::docstring::DocStrings, DocStyle, Renderable,
+    },
+    RenderPlan,
+};
+use anyhow::{anyhow, Result};
+use horrorshow::{box_html, Raw, RenderBox, Template};
+use std::{collections::BTreeMap, fmt::Write};
+use sway_core::language::ty::{TyEnumVariant, TyStorageField, TyStructField, TyTraitFn};
+
+/// The actual context of the item displayed by [ItemContext].
+/// This uses [ContextType] to determine how to represent the context of an item.
+///
+/// Example:
+/// ```sw
+/// struct Foo {}
+/// trait Foo {
+///     fn foo() -> Foo;
+/// }
+/// ```
+/// Becomes:
+/// ```rust
+/// Context {
+///     module_info: ModuleInfo, /* cloned from item origin to create links */
+///     context_type: ContextType::RequiredMethods(Vec<TyTraitFn>), /* trait fn foo() stored here */
+/// }
+/// ```
+#[derive(Clone, Debug)]
+pub(crate) struct Context {
+    module_info: ModuleInfo,
+    context_type: ContextType,
+}
+impl Context {
+    pub(crate) fn new(module_info: ModuleInfo, context_type: ContextType) -> Self {
+        Self {
+            module_info,
+            context_type,
+        }
+    }
+}
+impl Renderable for Context {
+    fn render(self, render_plan: RenderPlan) -> Result<Box<dyn RenderBox>> {
+        let mut rendered_list: Vec<String> = Vec::new();
```

Comment:
Perhaps inside each match arm before the loop consider reallocating to the size of the field/variant/method so that the pushing does not continue to reallocate? This would be a single change vs multiple changes depending on the number of items

Reply:
Makes sense

Reply:
After looking at this a bit more, the size of the `String` created in each iteration cannot be known until it's created. The field/variant/methods are not the same size as the `String` created from them.